keps/sig-network/2091-cluster-network-policy/README.md

```diff
@@ -0,0 +1,1161 @@
+
+(TODO) In the multi-cluster context, we are thinking to name the extended NetworkPolicy
+to MultiClusterNetworkPolicy. To avoid the extension of this API resource to be named
+MultiCluster-ClusterNetworkPolicy, maybe we should name this resource differently
+to begin with? AdminNetworkPolicy is a popular alternative.
+
+## Motivation
+
+Kubernetes provides NetworkPolicy resources to control traffic within a
+cluster. NetworkPolicies focus on expressing a developers intent to secure
+their applications. Thus, in order to satisfy the needs of a security admin,
+we propose to introduce a new API that captures the administrators intent.
+
+### Goals
+
+The goals for this KEP are to satisfy the following key user stories:
+1. As a cluster administrator, I want to enforce irrevocable guardrails that
+   all workloads must adhere to in order to guarantee the safety of my clusters.
+   In particular I want to enforce certain network level access controls that are
+cluster scoped and cannot be overridden or bypassed by namespace scoped network policies.
+2. As a cluster administrator, I want to have the option to enforce network level access controls
+   that facilitate network multi-tenancy and strict network level isolation between multiple
+   teams and tenants sharing a cluster via use of namespaces or groupings of namespaces per tenant.
+3. As a cluster administrator, I want to optionally also deploy an additional default set of policies to all
+   workloads that may be overridden by the developers if needed.
+
+There are several unique properties that we need to add in order accomplish the
+user stories above.
+1. Deny rules and, therefore, hierarchical enforcement of policy
+2. Semantics for a cluster-scoped policy object that may include
+   namespaces/workloads that have not been created yet.
+3. Backwards compatibility with existing Kubernetes Network Policy API
+
+### Non-Goals
+
+Our mission is to solve the most common use cases that cluster admins have.
+That is, we don't want to solve for every possible policy permutation a user
+can think of.  Instead, we want to design an API that addresses 90-95% use cases
+while keeping the mental model easy to understand and use.
+The focus of this KEP is on cluster scoped controls for east-west traffic within
+a cluster. Cluster scoped controls for North-South traffic may be addressed via
+future versions of the api resources introduced in this or other future KEPs.
+
+Additionally, this proposal is squarely focused on solving the needs of the
+Cluster Administrator. It is not intended to solve:
+1. Error reporting for Network Policy
+2. Kubernetes Node policies
+3. New policy selectors (services, service accounts, etc.)
+4. Cluster external traffic access control.
+
+## Proposal
+
+In order to achieve the two primary broad use cases for a cluster admin to
+secure K8s clusters, we propose to introduce the following resource
+under `netpol.networking.k8s.io` API group:
+- ClusterNetworkPolicy
```

Comment:
There was some discussion on this, but I think the WG is pretty flexible as it pertains to naming.  CC @abhiraut @srampal @Dyanngg

Reply:
this was a placeholder .. yes.. `policy` sounds good... there were other suggestions too.. I think `security` was one?

Reply:
So is the next feature to come along going to run into this same problem? (Having to come up with a new group name because it can't reuse any existing group name?) Or is this something that now that we know the problem exists, we can do things differently to avoid killing off `v1beta1` in the new group?

Anyway, if this group is going to end up being for CNP only and nothing else, then it should have a more "specific" name (like `policy`) rather than using up a more "generic" name (like `security`) that might be better used for some other feature later.

Reply:
I thing we should try to keep v1beta1 alive for the `policy` group, it doesn't make sense to have a new group for every new resource that comes out of the sig-network-policy API subgroup

Reply:
I don't remember the details... this was discussed a while back (either in the context of CNP or Ingress v2), and someone said that because of the way past API deprecations had been handled, that `networking.k8s.io/v1beta1` cannot be brought back.